File: frontend/src/types/expense.ts
/** 严格对齐后端 ExpenseStatus 七态 */
export enum ExpenseStatus {
  DRAFT = 'draft',
  SUBMITTED = 'submitted',
  PENDING = 'pending',
  APPROVED = 'approved',
  REJECTED = 'rejected',
  PAID = 'paid',
  CANCELLED = 'cancelled'
}

/** 严格对齐后端 ExpenseType 八值 */
export enum ExpenseType {
  TRAVEL = 'travel',
  OFFICE = 'office',
  ENTERTAINMENT = 'entertainment',
  TRAINING = 'training',
  MEAL = 'meal',
  TRANSPORT = 'transport',
  EQUIPMENT = 'equipment',
  OTHER = 'other'
}

export const ExpenseStatusLabels: Record<string, string> = {
  draft: '草稿',
  submitted: '已提交',
  pending: '待审批',
  approved: '已通过',
  rejected: '已驳回',
  paid: '已支付',
  cancelled: '已取消'
}

export const ExpenseStatusColors: Record<string, string> = {
  draft: 'info',
  submitted: 'primary',
  pending: 'warning',
  approved: 'success',
  rejected: 'danger',
  paid: 'success',
  cancelled: 'info'
}

export const ExpenseTypeLabels: Record<string, string> = {
  travel: '差旅费',
  office: '办公费',
  entertainment: '招待费',
  training: '培训费',
  meal: '餐饮费',
  transport: '交通费',
  equipment: '设备费',
  other: '其他'
}

/** 对齐后端 ExpenseItemResponse */
export interface ExpenseItem {
  id: number
  expense_id: number
  category_id?: number | null
  description: string
  amount: number
  expense_date: string
  invoice_no?: string | null
  invoice_url?: string | null
  invoice_verified?: boolean
  created_at: string
}

export interface ExpenseItemCreate {
  category_id?: number
  description: string
  amount: number
  expense_date: string
  /** 后端 ExpenseItemCreate.invoice_no（Optional[str], max_length=100），创建时可提交 */
  invoice_no?: string
  invoice_url?: string
}

export interface AIReviewIssue {
  type: string
  severity: 'low' | 'medium' | 'high'
  description: string
  item_index?: number
}

export interface AIReviewResult {
  risk_level: 'low' | 'medium' | 'high' | 'critical'
  risk_score: number
  /** 后端可能返回结构化对象而非字符串 */
  summary: string | Record<string, any>
  issues?: AIReviewIssue[]
  suggestions?: string[]
  reviewed_at?: string
}

/** 审批记录状态（对齐后端 ApprovalStatus） */
export type ApprovalStatus = 'pending' | 'approved' | 'rejected' | 'returned'

export const ApprovalStatusLabels: Record<string, string> = {
  pending: '待审',
  approved: '通过',
  rejected: '驳回',
  returned: '退回'
}

export const ApprovalStatusColors: Record<string, string> = {
  pending: 'info',
  approved: 'success',
  rejected: 'danger',
  returned: 'warning'
}

/**
 * 对齐后端 ApprovalResponse。
 * 注意：后端**没有** `action` 字段，审批动作体现在 `status` 上。
 * expense_no / title / expense_type / total_amount / expense_status 是列表接口
 * 附带的报销单冗余字段，详情/历史接口可能不返回，故全部可选。
 */
export interface ApprovalRecord {
  id: number
  expense_id: number
  approver_id: number
  flow_id?: number | null
  status: ApprovalStatus
  level: number
  comment?: string | null
  approver_name?: string | null
  expense_no?: string
  title?: string
  expense_type?: string
  total_amount?: number
  expense_status?: string
  created_at?: string
  updated_at?: string
}

/** 严格对齐后端 ExpenseResponse */
export interface Expense {
  id: number
  user_id: number
  expense_no: string
  title: string
  expense_type: ExpenseType | string
  total_amount: number
  currency: string
  status: ExpenseStatus | string
  description: string
  remark?: string | null
  submitted_at?: string | null
  approved_at?: string | null
  paid_at?: string | null
  risk_level?: string | null
  risk_score?: number | null
  ai_review_result?: AIReviewResult | Record<string, any> | null
  items: ExpenseItem[]
  created_at: string
  updated_at: string
}

export interface ExpenseCreate {
  title: string
  expense_type: string
  description: string
  items: ExpenseItemCreate[]
}

export interface ExpenseUpdate {
  title?: string
  expense_type?: string
  description?: string
}

export interface ExpenseListParams {
  page?: number
  page_size?: number
  status?: ExpenseStatus | string
  expense_type?: string
  keyword?: string
  sort_by?: string
  sort_order?: string
}

export interface AIReviewResponse {
  risk_level: 'low' | 'medium' | 'high' | 'critical'
  risk_score: number
  summary: string | Record<string, any>
  issues?: AIReviewIssue[]
  suggestions?: string[]
}

/** by_status / by_type 的桶结构 */
export interface StatisticsBucket {
  count: number
  amount: number
}

/**
 * GET /expenses/statistics/overview 的**真实** payload 形状。
 * 与下面的视图模型 `Statistics` 不同：后端返回的是 total_count + 两个按维度聚合的字典，
 * 由页面自行映射为视图模型。
 */
export interface StatisticsOverview {
  total_count: number
  total_amount: number
  by_status: Record<string, StatisticsBucket>
  by_type: Record<string, StatisticsBucket>
}

/** 页面侧的统计视图模型（由 StatisticsOverview 映射而来，非接口原始形状） */
export interface Statistics {
  total_expenses: number
  total_amount: number
  pending_count: number
  approved_count: number
  rejected_count: number
  draft_count: number
  monthly_stats: MonthlyStat[]
  type_stats: TypeStat[]
  department_stats: DepartmentStat[]
}

export interface MonthlyStat {
  month: string
  count: number
  total_amount: number
}

export interface TypeStat {
  type: string
  count: number
  total_amount: number
}

export interface DepartmentStat {
  department: string
  count: number
  total_amount: number
}

File: backend/app/schemas/expense.py
"""
报销单相关 Pydantic Schemas
"""
from datetime import datetime, date
from typing import Optional, List
from pydantic import BaseModel, Field
from decimal import Decimal


# ========== 费用明细 ==========

class ExpenseItemCreate(BaseModel):
    """费用明细创建请求"""
    category_id: Optional[int] = Field(None, description="费用类别ID")
    description: str = Field(..., min_length=1, max_length=500, description="费用说明")
    amount: float = Field(..., gt=0, description="金额")
    expense_date: Optional[date] = Field(None, description="费用发生日期")
    invoice_no: Optional[str] = Field(None, max_length=100, description="发票号码")
    invoice_url: Optional[str] = Field(None, max_length=500, description="发票URL")


class ExpenseItemResponse(BaseModel):
    """费用明细响应"""
    id: int
    expense_id: int
    category_id: Optional[int] = None
    description: str
    amount: float
    expense_date: Optional[date] = None
    invoice_no: Optional[str] = None
    invoice_url: Optional[str] = None
    invoice_verified: bool = False
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# ========== 报销单 ==========

class ExpenseCreate(BaseModel):
    """报销单创建请求"""
    title: str = Field(..., min_length=1, max_length=200, description="报销单标题")
    expense_type: str = Field(..., description="报销类型: travel/office/entertainment/transport/meal/training/equipment/other")
    description: Optional[str] = Field(None, max_length=1000, description="报销说明")
    currency: str = Field(default="CNY", max_length=10, description="币种")
    items: List[ExpenseItemCreate] = Field(..., min_length=1, description="费用明细列表")


class ExpenseUpdate(BaseModel):
    """报销单更新请求"""
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    expense_type: Optional[str] = None
    description: Optional[str] = Field(None, max_length=1000)
    currency: Optional[str] = Field(None, max_length=10)
    remark: Optional[str] = Field(None, max_length=500)
    items: Optional[List[ExpenseItemCreate]] = None


class ExpenseResponse(BaseModel):
    """报销单响应"""
    id: int
    user_id: int
    expense_no: str
    title: str
    expense_type: str
    total_amount: float = 0.0
    currency: str = "CNY"
    status: str = "draft"
    description: Optional[str] = None
    remark: Optional[str] = None
    submitted_at: Optional[datetime] = None
    approved_at: Optional[datetime] = None
    paid_at: Optional[datetime] = None
    risk_level: Optional[str] = None
    risk_score: Optional[float] = None
    ai_review_result: Optional[dict] = None
    items: List[ExpenseItemResponse] = []
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ExpenseListResponse(BaseModel):
    """报销单列表响应（带分页）"""
    success: bool = True
    data: List[ExpenseResponse] = []
    total: int = 0
    page: int = 1
    page_size: int = 20
    total_pages: int = 0


class ExpenseDetailResponse(BaseModel):
    """报销单详情响应"""
    success: bool = True
    data: ExpenseResponse
    message: str = "获取成功"


# ========== AI 审核 ==========

class AIReviewRequest(BaseModel):
    """AI审核请求"""
    expense_id: int = Field(..., description="报销单ID")
    include_history: bool = Field(default=False, description="是否包含历史记录分析")
    custom_rules: Optional[dict] = Field(None, description="自定义审核规则")


class AIReviewResponse(BaseModel):
    """AI审核响应"""
    success: bool = True
    data: dict = Field(default_factory=dict, description="AI审核结果详情")
    message: str = "AI审核完成"


# ========== 统计 ==========

class StatisticsResponse(BaseModel):
    """报销统计响应"""
    success: bool = True
    data: dict = Field(default_factory=dict)
    message: str = "获取成功"


# ========== 通用 ==========

class MessageResponse(BaseModel):
    """通用消息响应"""
    success: bool = True
    data: Optional[dict] = None
    message: str = "操作成功"
